Partially revert commit 9668d67

mikrosk opened the pull request:
The commit in question assumed that the only use of O_GLOBAL is inside
slip.xif. However this is not true: it was used in both N.AES and AES
4.1 (and perhaps others) as a way of accessing the console device from
AES by every process.

Roughly, it works like this in AES 4.1:

- "MiNT" (the kernel) opens /dev/console for read/write and stores the
  file handle as CON:

- "AESSYS" opens /dev/console for read/write with O_GLOBAL and gets a
  file handle as if it were "MiNT"

- ("AESSYS" somehow distributes returned file handle to everyone
  interested, perhaps as "this is the console device")

- everyone can now read and write from/to that file handle as if they
  were all root processes (i.e. "AESSYS" is treated as "MiNT" and so are
  others like "NEWDESK" or "SCREEN")

Narrow the O_GLOBAL handling only to /dev/console and allow max 32k
opened files per process (currently the limit is set to 1024).

Fixes #143.
Fixes #173.
Fixes #378.

mikrosk commented:
Btw, Helmut's change in ebbc3f33 does not remove the `O_GLOBAL` flag, just ignore it but I think his original proposed patch is cleaner.

mikrosk commented:
Fixes also #378 & #173.

th-otto commented:
The `ALERT` should also be removed. Function like Fopen() should just return an error code, not display an alert box on error.

mikrosk commented:
Detection is easy (comparing `name` to `/dev/console`) however I'm worried about the double-click issue with AES 4.1 -- that seems to need the whole O_GLOBAL deal, not just the workaround. I'll take a deeper look.

mikrosk commented:
Yeah, it's because AES 4.1 calls `sys_f_instat` heavily, probably missing some events in the process. I have posted my idea to freemint-discuss for further approval.

mikrosk commented:
I have updated the branch with a brand new implementation (and updated the description in this PR). It's basically the simplest way. Forcing some specific file handle values to be reserved for `/dev/console` doesn't seem so good after all as this would limit the number of processes which are allowed to open `/dev/console` directly to one.

mikrosk commented:
I tried to implement @th-otto's suggestion of using FILEPTR->flags. Unfortunately, it wont work, same as my idea of having a dedicated handle wont work. Even if we assume that only one process (AESSYS) opens /dev/console.

The main issue is that we still need some kind of mapping.

If we let /dev/console to be opened as the root process ("MiNT"), we need to map the root process' file handle back to any process accessing /dev/console using that handle (i.e. subtracting 100 from the handle value as done in the original code).

If we let /dev/console to be opened as usual (i.e. by "AESSYS"), we need to map *that* process' file handle back to any *other* process which is accessing /dev/console using that handle. So not only we need to store the handle value, we also need to store the process ID of "AESSYS" which opened /dev/console.

In neither case we can use a flag in FILEPTR, because we get the FILEPTR pointer using an index represented by the file handle. So the file handle must already contain some hint _which_ FILEPTR should be used.

So I propose to merge this PR as is, it really is the simplest solution and it works well even with multiple Fopen("/dev/console") calls. It could even support SLIP.XIF in the same fashion. And it has been verified as working by another user in https://github.com/freemint/freemint/issues/173#issuecomment-2355164031.

th-otto commented:
On Samstag, 21. September 2024 16:16:36 CEST Miro Kropáček wrote:
> In neither case we can use a flag in FILEPTR, because we get the FILEPTR
> pointer using an index represented by the file handle

Shouldn't something like this work?

In fp_get(), adding something like

if ((*fp)->flags & O_GLOBAL)
   *fp = rootproc->console_handle

(and similar in fp_get1)

Then in sys_f_open, make an entry of the FILEPTR both to the process file 
handle table, mark that with O_GLOBAL, and store that also to rootproc-
>console_handle.


That should work atleast for /dev/console (but not for others since pre-
process handles may be different)

mikrosk commented:
> Shouldn't something like this work?
>
> if ((*fp)->flags & O_GLOBAL)
>   *fp = rootproc->console_handle

No, because `fp` is uninitialised at this point, you can't use it for the check against O_GLOBAL. Its initialisation is a bit hidden a few lines below: `!(*fp = (*p)->p_fd->ofiles[*fd])`.

th-otto commented:
Yes, i know, i meant to add those line at the *end* of the function.

mikrosk commented:
> Yes, i know, i meant to add those line at the end of the function.

Wouldn't this corrupt some internal structures related to /dev/console? I mean, the first user of /dev/console is the root process which opens it and stores as file handle 0. Then AESSYS opens it, too (with O_GLOBAL) and in the original implementation it gets _another_ root process handle (usually 6). Now AESSYS would get 0. So what if those two processes read from / write to /dev/console in parallel? I know it's not a regular file but still, some internal per-file and per-process values may get confused?

mikrosk commented:
Actually, I had a similar code done and now I remembered why it didn't work: console's functions like `sys_f_instat` and `sys_f_getchar` do not rely on previous Fopen("/dev/console"), that's the whole point of O_GLOBAL. So AESSYS opens /dev/console ... gets file handle 8 ... and then later, process "NEWDESK" comes and starts accessing stuff via that file handle = 8 which is of course bogus.

th-otto commented:
Oh, ok. So a *different* process uses the same handle, that won't  work of course with my approach. I just wonder: how is the handle passed from N.AES to NEWDESK? Just by keeping it open when AES executes it? If that is the case, then the approach could still work. You just have to adopt the code that inherits the handles from the parent.

mikrosk commented:
>  I just wonder: how is the handle passed from N.AES to NEWDESK? Just by keeping it open when AES executes it? If that is the case, then the approach could still work. You just have to adopt the code that inherits the handles from the parent.

Frankly, I don't know. I mentioned it in the first post, AES (I'm testing the 4.1 version by Atari, not N.AES) somehow tells everyone ("SCREEN", "NEWDESK", ...) which handle they are supposed to use for accessing the console. And then they do, bypassing Fopen("/dev/console") of their own. Why they can't just open /dev/console? No clue...

mikrosk commented:
I couldn't come up with anything smarter and since it fixes the problem, I'm merging it. In the future we can rework this again, now we are at least aware of all the side effects and bumps along the round.

Oh and as a bonus, we get slip.xif again. Not that I expect anyone testing it anytime soon but it's there, usable for anyone interested.

th-otto commented:
I'm not really happy with it, since it is an ugly hack. It could even imagine that some people now think they can use O_GLOBAL again.

>Oh and as a bonus, we get slip.xif again.

Not really, since it is not even compiled currently.

mikrosk commented:
Thorsten, sometimes I really wish you verified facts before writing your replies.

1. The code is limited to opening "u:\dev\console" and "u:\pipe\sld" as this is the only use case which makes sense for O_GLOBAL. I'm not sure who else could make some use of those.

2. This PR enabled slip.xif building, it's already in the snapshots.

czietz commented:
If someone is very keen to test, this device uses SLIP: https://www.chzsoft.de/site/hardware/connecting-the-atari-st-to-the-internet/.

(But I have ripped the hardware apart to use it in other projects; so I won't test.)
